### PR TITLE
htmldocck: Add support for `/text()` in `@snapshot`

### PR DESCRIPTION
**Pull request by camelid**
This allows just testing the text, in cases where the HTML tags don't
matter.

See https://github.com/rust-lang/rust/pull/92908#discussion_r785191758 for an example of when this would be useful.

r? @GuillaumeGomez

### CODE REVIEW

**Comment by camelid** [comment]
cc @jsha: I think this should mitigate your concern about `@snapshot` leading to brittle tests?

File: src/etc/htmldocck.py [hunk 1]
```diff
@@ -494,11 +497,16 @@ def check_command(c, cache):
                 [snapshot_name, html_path, pattern] = c.args
                 tree = cache.get_tree(html_path)
                 xpath = normalize_xpath(pattern)
+                if xpath.endswith('/text()'):
+                    xpath = xpath[:-7]
+                    normalize_to_text = True
```
Comment (camelid) [inline src/etc/htmldocck.py:502]: This duplicates code elsewhere in htmldocck.

**Comment by jsha** [comment]
How does this differ from our current XPath checks, where the stuff between the quotes is a text match? E.g.:

```
// @has cap_lints/index.html //* 'Crate cap_lints'
// @has cap_lints/struct.Foo.html //* 'Foo'
```

Also, is `/text()` valid XPath? Or does this move us further from a standard XPath deployment?

**Comment by camelid** [comment]
> How does this differ from our current XPath checks, where the stuff between the quotes is a text match? E.g.:

It makes it more readable and enables diffing and `bless`ing (rather than just saying "XPath not found").

> Also, is `/text()` valid XPath? Or does this move us further from a standard XPath deployment?

Yes, `/text()` is valid XPath. Also, `/text()` is already supported by `@has`, in a similarly hacky way.

**Comment by camelid** [comment]
An example of how the old way is not very readable:

```rust
// @has macro_generated_macro/macro.linebreak.html //pre 'macro_rules! linebreak {'
// @has - //pre '    ('
// @has - //pre '        <= 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25'
// @has - //pre '        26 27 28 =>'
// @has - //pre '    ) => { ... };'
// @has - //pre '};'
```

from an old version of #92908.

File: src/etc/htmldocck.py [hunk 2]
```diff
@@ -494,11 +497,16 @@ def check_command(c, cache):
...
+                else:
+                    normalize_to_text = False
```
Comment (GuillaumeGomez) [inline src/etc/htmldocck.py:504]:
```suggestion
            normalize_to_text = False
            if xpath.endswith('/text()'):
                    xpath = xpath[:-7]
                    normalize_to_text = True
```

**Comment by GuillaumeGomez** [comment]
Yes, that seems good to me.

On a somewhat similar topic: it'd be nice that we merged https://github.com/rust-lang/rust/pull/89676 at some point to really support XPath completely... Currently it's very hacky and pretty bad.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
I'm fine with this dd. Could you maybe a test making use of it please?

**Comment by dtolnay** [comment]
Is it good enough that I'll be adding tests making use of it in #92908?

**Comment by GuillaumeGomez** [comment]
That sounds good to me!

@bors: r+ rollup

**Pull request merged**